cpp/Esri/ArcGISRuntime/Toolkit/GenericListModel.cpp

```diff
@@ -0,0 +1,379 @@
+    const auto property = m_elementType->property(propIndex);
+    return property.read(o);
+  }
+  else
+  {
+    return QVariant();
+  }
+}
+
+bool GenericListModel::setData(const QModelIndex& index, const QVariant& value, int role)
+{
+  if (!m_elementType)
+    return false;
+  else if (!index.isValid())
+    return false;
+
+  if (role == Qt::DisplayRole || role == Qt::EditRole)
+  {
+    auto o = m_objects.at(index.row());
+    const auto property = m_elementType->property(m_displayPropIndex);
+    return property.write(o, value);
+  }
+  else if (role == Qt::UserRole)
+  {
+    auto newObject = qvariant_cast<QObject*>(value);
+    if (newObject && m_elementType == newObject->metaObject())
+    {
+      m_objects[index.row()] = newObject;
+      emit dataChanged(index, index);
+      return true;
+    }
+    else
+    {
+      return false;
+    }
+  }
+  else if (role > Qt::UserRole)
+  {
+    auto o = m_objects.at(index.row());
+    const auto propIndex = role - (Qt::UserRole + 1) + m_elementType->propertyOffset();
+    const auto property = m_elementType->property(propIndex);
+    return property.write(o, value);
+  }
+  else
+  {
+    return false;
+  }
+}
+
+QHash<int, QByteArray> GenericListModel::roleNames() const
+{
+  if (!m_elementType)
+    return { };
+
+  QHash<int, QByteArray> output;
+  output.insert( Qt::UserRole, "modelData");
+
+  const int offset = m_elementType->propertyOffset();
+  for (int i = offset; i < m_elementType->propertyCount(); ++i)
+  {
+    auto property = m_elementType->property(i);
+    output.insert(i - offset + Qt::UserRole + 1, property.name());
+  }
+  return output;
+}
+
+bool GenericListModel::insertRows(int row, int count, const QModelIndex& parent)
+{
+  if (!m_elementType)
+    return false;
+  if (parent.isValid())
+    return false;
+  else if (count < 1)
+    return false;
+  else if (row < 0)
+    return false;
+  else if (row > m_objects.size())
+    return false;
+
+  beginInsertRows(parent, row, row + count - 1);
+  for (int i = 0; i < count; ++i)
+  {
+    m_objects.insert(row, m_elementType->newInstance(Q_ARG(QObject*, this)));
+  }
+  endInsertRows();
+
+  for (int i = row; i < count; ++i)
+  {
+    connectElement(index(row));
+  }
+  return true;
+}
+
+bool GenericListModel::removeRows(int row, int count, const QModelIndex& parent)
+{
+  if (parent.isValid())
+    return false;
+  else if (count < 0)
+    return false;
+  else if (row < 0)
+    return false;
+  else if (row >= rowCount(parent))
+    return false;
+  else if (row + count > rowCount(parent))
+    return false;
+  else if (count == 0)
+    return true; // A valid no-op.
+
+  beginRemoveRows(parent, row, row + count - 1);
+  for (int i = count - 1; i >= row; --i)
+  {
+    auto o = m_objects.at(i);
+
+    // Ensure additional removal signals are not triggered.
+    if (o)
+      disconnect(o, &QObject::destroyed, this, nullptr);
+
+    delete o;
```

Comment:
Should we be storing raw pointers if we need to do this? This seems like a red flag.

Reply:
We listen into `&QObject::destroyed` signals in the `connectElement` method, so by the time we reach here the pointer will be null if already deleted outside of this model.

What are you suggesting? A list of `QPoint<QObject>` instead? That might actually be quite a bit cleaner than the current system.

Reply:
My comment on this is that it makes me nervous to see raw delete calls. I realize shared pointers probably won't work, but I wonder if there is any other solution to avoid this?